File: app/src/main/java/com/jameskbride/eventbusmvpdemo/network/NetworkApiWrapper.java
package com.jameskbride.eventbusmvpdemo.network;

import com.jameskbride.eventbusmvpdemo.bus.NetworkErrorEvent;
import com.jameskbride.eventbusmvpdemo.bus.NetworkRequestEvent;

import org.greenrobot.eventbus.EventBus;

import java.io.IOException;

import io.reactivex.functions.Consumer;

public class NetworkApiWrapper<T extends Throwable> implements Consumer<T> {

    private EventBus eventBus;
    private Consumer consumer;
    private NetworkRequestEvent networkRequestEvent;

    public NetworkApiWrapper(EventBus eventBus, Consumer consumer, NetworkRequestEvent networkRequestEvent) {
        this.eventBus = eventBus;
        this.consumer = consumer;
        this.networkRequestEvent = networkRequestEvent;
    }

    @Override
    public void accept(T t) throws Exception {
        if (t instanceof IOException) {
            eventBus.post(new NetworkErrorEvent(networkRequestEvent));
        } else {
            consumer.accept(t);
        }
    }
}

File: app/src/main/java/com/jameskbride/eventbusmvpdemo/network/NetworkApiWrapper.kt
package com.jameskbride.eventbusmvpdemo.network

import com.jameskbride.eventbusmvpdemo.bus.BusAware
import com.jameskbride.eventbusmvpdemo.bus.NetworkErrorEvent
import com.jameskbride.eventbusmvpdemo.bus.NetworkRequestEvent
import io.reactivex.functions.Consumer
import org.greenrobot.eventbus.EventBus
import java.io.IOException


class NetworkApiWrapper<T : Throwable>(
        override val eventBus: EventBus,
        val consumer:Consumer<T>,
        val networkRequestEvent: NetworkRequestEvent) : Consumer<T>, BusAware {

    override fun accept(t: T) {
        if (t is IOException) {
            eventBus.post(NetworkErrorEvent(networkRequestEvent))
        } else {
            consumer.accept(t)
        }
    }
}